Corrected _temp_top.top behaviour to accommodate asynchronous scenarios

ale94mleon opened the pull request:
## Issue Addressed:

Encountered a `FileNotFoundError: [Errno 2] No such file or directory: '.../_temp_top.top'` error during asynchronous execution of `gmx_MMPBSA`, particularly when multiple jobs sharing the same topology file were run concurrently (e.g., using replicas).

## Solution Overview:

The issue stemmed from one job deleting the topology file before another job could access it. To mitigate this, the fix involved utilizing `tempfile.NamedTemporaryFile(dir=top_file.parent, prefix='_temp_top', suffix='.top', mode='w')`. This approach ensures the creation of a unique temporary topology file for each parallel job instance. Notably, the temporary file is generated in the directory of `top_file` to prevent recurrence of issue #299.

Valdes-Tresanco-MS commented:
Hey @ale94mleon
Thank you for your contribution.
This PR doesn't work for me.

Using `NamedTemporaryFile` in context keeps the file in writing mode, so it is empty until the context ends and it is closed. When you read the file with parmed inside the context you get an error as expected, because the file is in writing mode yet.
The solution is setting `delete=False` and reading the NamedTemporaryFile outside the context. No matter that the temp_top is kept in the folder

ale94mleon commented:
Hum, strange, I am not getting this issue. with this minimalist example

```python
import tempfile
from pathlib import Path
import parmed


top_file = Path('complex.top').resolve()
molsect = False


with tempfile.NamedTemporaryFile(dir=top_file.parent, prefix='_temp_top', suffix='.top', mode='w') as temp_top:
    # temp_top.write('; Modified by gmx_MMPBSA\n')
    # TODO: keep solvent when n-wat is implemented
    with open(top_file) as topf:
        for line in topf:
            if '[ molecules ]' in line:
                molsect = True
            if molsect:
                # not copy ions and solvent
                sol_ion = [
                    # standard gmx form
                    'NA', 'CL', 'SOL', 'K'
                    # charmm-GUI form ??
                                    'SOD', 'Na+', 'CLA', 'Cl-', 'POT', 'K+',
                    'TIP3P', 'TIP3', 'TP3', 'TIPS3P', 'TIP3o',
                    'TIP4P', 'TIP4PEW', 'T4E', 'TIP4PD',
                    'TIP5P',
                    'SPC', 'SPC/E', 'SPCE',
                    'WAT',
                    'OPC']

                if not line.split():
                    continue
                if line.split()[0].strip() in sol_ion:
                    continue
            temp_top.write(line)

    # read the temp topology with parmed
    rtemp_top = parmed.gromacs.GromacsTopologyFile(temp_top.name)

print(rtemp_top.atoms)
```

I am getting:

```
AtomList([
        <Atom C1 [0]; In WP6 0>
        <Atom C2 [1]; In WP6 0>
        <Atom C3 [2]; In WP6 0>
        <Atom C4 [3]; In WP6 0>
        <Atom C5 [4]; In WP6 0>
        <Atom C6 [5]; In WP6 0>
        <Atom C7 [6]; In WP6 0>
        <Atom O1 [7]; In WP6 0>
        <Atom O2 [8]; In WP6 0>
        <Atom C8 [9]; In WP6 0>
        <Atom C9 [10]; In WP6 0>
        <Atom C10 [11]; In WP6 0>
        <Atom C11 [12]; In WP6 0>
        <Atom O3 [13]; In WP6 0>
        <Atom O4 [14]; In WP6 0>
        <Atom O5 [15]; In WP6 0>
        <Atom O6 [16]; In WP6 0>
        <Atom C12 [17]; In WP6 0>
        <Atom C13 [18]; In WP6 0>
        <Atom C14 [19]; In WP6 0>
        <Atom C15 [20]; In WP6 0>
        <Atom C16 [21]; In WP6 0>
        <Atom C17 [22]; In WP6 0>
        <Atom C18 [23]; In WP6 0>
        ...
        <Atom H22 [184]; In LIG 1>
        <Atom H23 [185]; In LIG 1>
        <Atom H24 [186]; In LIG 1>
        <Atom H25 [187]; In LIG 1>
        <Atom H26 [188]; In LIG 1>
])
```

So, `parmed-4.2.2` (in macOS) is actually reading the content, but I get your point in the writing mode. What version are you using of `parmed`? Could you please try the previous example but with `mode='w+'`. The topology is attached.
[topology.zip](https://github.com/Valdes-Tresanco-MS/gmx_MMPBSA/files/15180294/topology.zip)

Valdes-Tresanco-MS commented:
The problem is in the topology file... In your example, it works, but when the topology has itp references fails
For example the [Alanine_scanning](https://valdes-tresanco-ms.github.io/gmx_MMPBSA/dev/examples/Alanine_scanning/)
```
;;
;; Generated by CHARMM-GUI FF-Converter
;;
;; Correspondance:
;; jul316@lehigh.edu or wonpil@lehigh.edu
;;
;; The main GROMACS topology file
;;

; Include forcefield parameters
#include "toppar/forcefield.itp"
#include "toppar/PROA.itp"
#include "toppar/PROB.itp"
#include "toppar/Na+.itp"
#include "toppar/Cl-.itp"
#include "toppar/TP3.itp"

[ system ]
; Name
Title

[ molecules ]
; Compound	#mols
PROA  	           1
PROB  	           1
Na+   	          11
Cl-   	          10
TP3   	        3608
```

Curiously, everything works well before the reading :(
Please, check if you get the same behavior

ale94mleon commented:
This is a really funny case. The following code works in the [Alanine_scanning](https://valdes-tresanco-ms.github.io/gmx_MMPBSA/dev/examples/Alanine_scanning/) ([ef4bf3e](https://github.com/Valdes-Tresanco-MS/gmx_MMPBSA/pull/493/commits/ef4bf3e4ac7663158b0acdc49800577ac0bb0b64)). But I still do not get why in the first example works and in the topology with include statements does not :-/

```python
import tempfile
from pathlib import Path
import parmed


top_file = Path('topol.top').resolve()
molsect = False


with tempfile.NamedTemporaryFile(dir=top_file.parent, prefix='_temp_top_', suffix='.top', mode='w', delete=False) as temp_top:
    # temp_top.write('; Modified by gmx_MMPBSA\n')
    # TODO: keep solvent when n-wat is implemented
    with open(top_file) as topf:
        for line in topf:
            if '[ molecules ]' in line:
                molsect = True
            if molsect:
                # not copy ions and solvent
                sol_ion = [
                    # standard gmx form
                    'NA', 'CL', 'SOL', 'K'
                    # charmm-GUI form ??
                                    'SOD', 'Na+', 'CLA', 'Cl-', 'POT', 'K+',
                    'TIP3P', 'TIP3', 'TP3', 'TIPS3P', 'TIP3o',
                    'TIP4P', 'TIP4PEW', 'T4E', 'TIP4PD',
                    'TIP5P',
                    'SPC', 'SPC/E', 'SPCE',
                    'WAT',
                    'OPC']

                if not line.split():
                    continue
                if line.split()[0].strip() in sol_ion:
                    continue
            temp_top.write(line)

# read the temp topology with parmed
rtemp_top = parmed.gromacs.GromacsTopologyFile(temp_top.name)

print(rtemp_top.atoms)

Path(temp_top.name).unlink()
```

As you mentioned `delete=False` do the trick, and after the closing of the temporal file, you can just treat it as a normal file.

Valdes-Tresanco-MS commented:
Yeah, this is fun.
It works if you set `mode="w+` and then add `temp_top.read()` after the `topf` context. lol.
I think this solution does the trick
